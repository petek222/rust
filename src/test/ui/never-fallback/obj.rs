#![feature(never_type)]
#![feature(never_type_fallback)]

fn get_type<T>(_: T) -> &'static str {
    std::any::type_name::<T>()
}

fn unconstrained_return<T>() -> Result<T, String> {
    Err("Hi".to_string())
}

fn foo() {
    let a = || {
        match unconstrained_return::<_>() { //~ ERROR Fallback to `!` may introduce undefined
            Ok(x) => x,  // `x` has type `_`, which is unconstrained
            Err(s) => panic!(s),  // â€¦ except for unifying with the type of `panic!()`
            // so that both `match` arms have the same type.
            // Therefore `_` resolves to `!` and we "return" an `Ok(!)` value.
        }
    };

    let cast: &dyn FnOnce() -> _ = &a;
    println!("Return type: {:?}", get_type(cast));
}

fn main() {
    foo()
}
