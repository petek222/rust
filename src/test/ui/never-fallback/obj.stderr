error: Fallback to `!` may introduce undefined behavior
  --> $DIR/obj.rs:14:15
   |
LL |         match unconstrained_return::<_>() {
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the type here was inferred to `!`
  --> $DIR/obj.rs:14:38
   |
LL |         match unconstrained_return::<_>() {
   |                                      ^
note: ... due to this expression evaluating to `!`
  --> $DIR/obj.rs:16:23
   |
LL |             Err(s) => panic!(s),  // â€¦ except for unifying with the type of `panic!()`
   |                       ^^^^^^^^^
   = note: If you want the `!` type to be used here, add explicit type annotations
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to previous error

